File: src/bin/lc-1584.rs
// use std::{collections::HashSet, vec};

struct Solution;

// later use Disjoint-Set structure (a bit awkward in rust... or is it?)
// struct DJSNode<'a> {
//     parent: Option<&'a DJSNode<'a>>,
// }

// impl<'a> DJSNode<'a> {
//     fn new() -> Self {
//         let mut node = DJSNode { parent: None };
//         node.find();
//         node
//     }

//     fn find(&mut self) -> &'a DJSNode {
//         let parent = match self.parent {
//             Some(p) => p.find(),
//             None => self,
//         };
//         self.parent = Some(parent);
//         parent
//     }

//     fn union(&mut self) {}

//     fn size(&self) {}
// }

impl Solution {
    pub fn min_cost_connect_points(points: Vec<Vec<i32>>) -> i32 {
        let mut edges = Vec::new();
        for i in 0..points.len() {
            for j in 0..i {
                let cost =
                    points[i][0].abs_diff(points[j][0]) + points[i][1].abs_diff(points[j][1]);
                edges.push((i, j, cost))
            }
        }
        edges.sort_by_key(|k| k.2);
        let mut cost = 0;
        let mut p: Vec<_> = (0..points.len()).collect();
        for (i, j, _cost) in edges {
            let pi = p[i];
            let pj = p[j];
            if p[i] != p[j] {
                cost += _cost;
                let pij = p[i].min(p[j]);
                // Probably expensive; without disjoint set. not -that- bad tho
                for _p in p.iter_mut() {
                    if *_p == pi || *_p == pj {
                        *_p = pij;
                    }
                }
            }
        }
        cost as i32
    }

    pub fn min_cost_connect_points_2(points: Vec<Vec<i32>>) -> i32 {
        let mut edges = Vec::new();
        for i in 0..points.len() {
            for j in 0..i {
                let cost =
                    points[i][0].abs_diff(points[j][0]) + points[i][1].abs_diff(points[j][1]);
                edges.push((i, j, cost))
            }
        }
        edges.sort_by_key(|k| k.2);
        let mut cost = 0;
        let mut p: Vec<_> = (0..points.len()).collect();
        for (i, j, _cost) in edges {
            let pi = p[i];
            let pj = p[j];
            if p[i] != p[j] {
                cost += _cost;
                let pij = p[i].min(p[j]);
                // Probably expensive; without disjoint set. not -that- bad tho
                for _p in p.iter_mut() {
                    if *_p == pi || *_p == pj {
                        *_p = pij;
                    }
                }
            }
        }
        cost as i32
    }
}

fn main() {
    for f in [
        Solution::min_cost_connect_points,
        Solution::min_cost_connect_points_2,
    ] {
        assert_eq!(
            f(vec![
                vec![0, 0],
                vec![2, 2],
                vec![3, 10],
                vec![5, 2],
                vec![7, 0]
            ]),
            20
        );
        assert_eq!(f(vec![vec![3, 12], vec![-2, 5], vec![-4, 1]]), 18);
        assert_eq!(
            f(vec![
                vec![2, -3],
                vec![-17, -8],
                vec![13, 8],
                vec![-17, -15]
            ]),
            53
        );
    }
}

File: cp/leetcode/lc-1584.py
import heapq


class DisjointSet:
    def __init__(self, value):
        self.value = value
        self.parent = self
        self.size = 1

    def find(self):
        root = self
        while root.parent != root:
            root = root.parent

        x = self
        while x.parent != root:
            parent = x.parent
            x.parent = root
            x = parent

        return root

    def union(self, other):
        x = self.find()
        y = other.find()

        if x != y:
            if x.size < y.size:
                x.parent = y
                y.size += x.size
            else:
                y.parent = x
                x.size += y.size

    def __str__(self):
        return f"{self.value} {self.size}"


class Solution:
    def minCostConnectPoints(self, points: list[list[int]]) -> int:
        edges = []
        for i in range(len(points)):
            for j in range(0, i):
                cost = abs(points[i][0] - points[j][0]) + abs(
                    points[i][1] - points[j][1]
                )
                edges.append((i, j, cost))

        # Join
        p = list(range(len(points)))
        edges.sort(key=lambda x: x[2])
        cost = 0
        for i, j, _cost in edges:
            pi = p[i]
            pj = p[j]
            if pi != pj:
                cost += _cost
                pij = min(pi, pj)
                for k in range(len(p)):
                    if p[k] == pi or p[k] == pj:
                        p[k] = pij
        return cost

    def minCostConnectPoints2(self, points: list[list[int]]) -> int:
        edges = []
        for i in range(len(points)):
            for j in range(0, i):
                cost = abs(points[i][0] - points[j][0]) + abs(
                    points[i][1] - points[j][1]
                )
                edges.append((i, j, cost))

        # Join
        edges.sort(key=lambda x: x[2])
        p = [DisjointSet(i) for i in range(len(points))]
        cost = 0
        for i, j, _cost in edges:
            pi = p[i].find()
            pj = p[j].find()
            if pi != pj:
                cost += _cost
                pi.union(pj)
        return cost

    def minCostConnectPoints3(self, points: list[list[int]]) -> int:
        adj_list = {}
        for i in range(len(points)):
            for j in range(0, i):
                cost = abs(points[i][0] - points[j][0]) + abs(
                    points[i][1] - points[j][1]
                )
                adj_list.setdefault(i, [])
                adj_list.setdefault(j, [])
                adj_list[i].append((cost, j))
                adj_list[j].append((cost, i))

        h = []
        for e in adj_list.get(0, []):
            heapq.heappush(h, e)
        f = {0}
        cost = 0
        while h:
            _cost, _v = heapq.heappop(h)
            if _v in f:
                continue
            f.add(_v)
            cost += _cost
            for e in adj_list[_v]:
                heapq.heappush(h, e)
        return cost


s = Solution()
for f in (s.minCostConnectPoints, s.minCostConnectPoints2, s.minCostConnectPoints3):
    assert f([[0, 0], [2, 2], [3, 10], [5, 2], [7, 0]]) == 20
    assert f([[3, 12], [-2, 5], [-4, 1]]) == 18
    assert f([[2, -3], [-17, -8], [13, 8], [-17, -15]]) == 53
    assert f([[0, 0]]) == 0
